server/src/main/java/org/elasticsearch/index/mapper/MapperService.java

```diff
@@ -365,24 +368,139 @@ public void merge(IndexMetadata indexMetadata, MergeReason reason) {
         }
     }
 
+    /**
+     * Merging the provided mappings. Actual merging is done in the raw, non-parsed, form of the mappings. This allows to do a proper bulk
+     * merge, where parsing is done only when all raw mapping settings are already merged.
+     */
+    public DocumentMapper merge(String type, List<CompressedXContent> mappingSources, MergeReason reason) {
+        final DocumentMapper currentMapper = this.mapper;
+        if (currentMapper != null && mappingSources.size() == 1 && currentMapper.mappingSource().equals(mappingSources.get(0))) {
+            return currentMapper;
+        }
+
+        Map<String, Object> mergedRawMapping = null;
+        for (CompressedXContent mappingSource : mappingSources) {
+            Map<String, Object> rawMapping = MappingParser.convertToMap(mappingSource);
+
+            // normalize mappings, making sure that all have the provided type as a single root
+            if (rawMapping.containsKey(type)) {
+                if (rawMapping.size() > 1) {
+                    throw new MapperParsingException("cannot merge a map with multiple roots, one of which is [" + type + "]");
+                }
+            } else {
+                rawMapping = Map.of(type, rawMapping);
+            }
+
+            if (mergedRawMapping == null) {
+                mergedRawMapping = rawMapping;
+            } else {
+                XContentHelper.merge(type, mergedRawMapping, rawMapping, RawFieldMappingMerge.INSTANCE);
+            }
+        }
+        if (mergedRawMapping != null && mergedRawMapping.size() > 1) {
+            throw new MapperParsingException("cannot merge mapping sources with different roots");
+        }
+        return (mergedRawMapping != null) ? doMerge(type, reason, mergedRawMapping) : null;
+    }
+
+    private static class RawFieldMappingMerge implements XContentHelper.CustomMerge {
```

Comment:
Can you add a javadoc for a high-level explanation of how this custom merger behaves? I.e., its special treatment of `subobjects`, etc.